File: risk/engine/combat.py
"""Combat resolution for Risk: single roll, attack execution, and blitz."""

import random

from pydantic import BaseModel

from risk.models.actions import AttackAction, BlitzAction
from risk.models.game_state import GameState, TerritoryState


class CombatResult(BaseModel):
    """Outcome of a single combat round."""

    attacker_losses: int
    defender_losses: int


def resolve_combat(
    attacker_dice: int, defender_dice: int, rng: random.Random
) -> CombatResult:
    """Roll dice and resolve a single round of combat.

    Dice are sorted descending and paired highest-first.
    Ties go to the defender.
    """
    attacker_rolls = sorted([rng.randint(1, 6) for _ in range(attacker_dice)], reverse=True)
    defender_rolls = sorted([rng.randint(1, 6) for _ in range(defender_dice)], reverse=True)

    attacker_losses = 0
    defender_losses = 0

    for a_die, d_die in zip(attacker_rolls, defender_rolls):
        if a_die > d_die:
            defender_losses += 1
        else:
            # Tie goes to defender
            attacker_losses += 1

    return CombatResult(attacker_losses=attacker_losses, defender_losses=defender_losses)


def validate_attack(
    state: GameState,
    map_graph: object,
    action: AttackAction,
    player_index: int,
) -> None:
    """Validate an attack action. Raises ValueError on invalid attack."""
    source_ts = state.territories[action.source]
    target_ts = state.territories[action.target]

    if source_ts.owner != player_index:
        raise ValueError(
            f"Player {player_index} does not own source territory '{action.source}'"
        )

    if target_ts.owner == player_index:
        raise ValueError(
            f"Cannot attack own/friendly territory '{action.target}'"
        )

    if not map_graph.are_adjacent(action.source, action.target):  # type: ignore[union-attr]
        raise ValueError(
            f"Territories '{action.source}' and '{action.target}' are not adjacent"
        )

    if source_ts.armies < action.num_dice + 1:
        raise ValueError(
            f"Source '{action.source}' has {source_ts.armies} armies but needs "
            f"at least {action.num_dice + 1} to attack with {action.num_dice} dice"
        )


def execute_attack(
    state: GameState,
    map_graph: object,
    action: AttackAction,
    player_index: int,
    rng: random.Random,
    defender_dice: int | None = None,
) -> tuple[GameState, CombatResult, bool]:
    """Execute a single attack round. Returns (new_state, result, conquered)."""
    validate_attack(state, map_graph, action, player_index)

    target_ts = state.territories[action.target]
    if defender_dice is None:
        defender_dice = min(2, target_ts.armies)

    result = resolve_combat(action.num_dice, defender_dice, rng)

    # Apply losses
    new_territories = dict(state.territories)
    source_ts = state.territories[action.source]

    new_source_armies = source_ts.armies - result.attacker_losses
    new_target_armies = target_ts.armies - result.defender_losses

    conquered = new_target_armies <= 0

    if conquered:
        # Territory conquered: transfer ownership, move armies
        armies_moved = action.num_dice
        new_source_armies -= armies_moved
        new_territories[action.source] = TerritoryState(
            owner=source_ts.owner, armies=new_source_armies
        )
        new_territories[action.target] = TerritoryState(
            owner=player_index, armies=armies_moved
        )
    else:
        new_territories[action.source] = TerritoryState(
            owner=source_ts.owner, armies=new_source_armies
        )
        new_territories[action.target] = TerritoryState(
            owner=target_ts.owner, armies=new_target_armies
        )

    update: dict = {"territories": new_territories}
    if conquered:
        update["conquered_this_turn"] = True

    new_state = state.model_copy(update=update)
    return new_state, result, conquered


def execute_blitz(
    state: GameState,
    map_graph: object,
    action: BlitzAction,
    player_index: int,
    rng: random.Random,
) -> tuple[GameState, list[CombatResult], bool]:
    """Auto-resolve attack: loop until conquest or attacker has 1 army."""
    # Validate basic preconditions
    source_ts = state.territories[action.source]
    target_ts = state.territories[action.target]

    if source_ts.owner != player_index:
        raise ValueError(
            f"Player {player_index} does not own source territory '{action.source}'"
        )
    if target_ts.owner == player_index:
        raise ValueError(
            f"Cannot attack own/friendly territory '{action.target}'"
        )
    if not map_graph.are_adjacent(action.source, action.target):  # type: ignore[union-attr]
        raise ValueError(
            f"Territories '{action.source}' and '{action.target}' are not adjacent"
        )
    if source_ts.armies < 2:
        raise ValueError(
            f"Source '{action.source}' needs at least 2 armies to attack"
        )

    current_state = state
    all_results: list[CombatResult] = []

    while True:
        src = current_state.territories[action.source]
        tgt = current_state.territories[action.target]

        attacker_dice = min(3, src.armies - 1)
        defender_dice = min(2, tgt.armies)

        if attacker_dice < 1:
            break

        attack_action = AttackAction(
            source=action.source, target=action.target, num_dice=attacker_dice
        )
        current_state, result, conquered = execute_attack(
            current_state, map_graph, attack_action, player_index, rng, defender_dice
        )
        all_results.append(result)

        if conquered:
            return current_state, all_results, True

        # Check if attacker reduced to 1
        if current_state.territories[action.source].armies <= 1:
            return current_state, all_results, False

    return current_state, all_results, False

File: mobile/lib/engine/combat.dart
import 'dart:math';

import 'actions.dart';
import 'map_graph.dart';
import 'models/game_state.dart';

/// Outcome of a single combat round.
class CombatResult {
  final int attackerLosses;
  final int defenderLosses;
  const CombatResult(
      {required this.attackerLosses, required this.defenderLosses});
}

/// Roll dice and resolve a single round of combat.
///
/// Dice are sorted descending and paired highest-first.
/// Ties go to the defender (attacker loses on tie).
CombatResult resolveCombat(
    int attackerDice, int defenderDice, Random rng) {
  final attackerRolls =
      List.generate(attackerDice, (_) => rng.nextInt(6) + 1)
        ..sort((a, b) => b.compareTo(a));
  final defenderRolls =
      List.generate(defenderDice, (_) => rng.nextInt(6) + 1)
        ..sort((a, b) => b.compareTo(a));

  int attackerLosses = 0;
  int defenderLosses = 0;

  final pairs = attackerRolls.length < defenderRolls.length
      ? attackerRolls.length
      : defenderRolls.length;

  for (int i = 0; i < pairs; i++) {
    if (attackerRolls[i] > defenderRolls[i]) {
      defenderLosses++;
    } else {
      // Tie goes to defender
      attackerLosses++;
    }
  }

  return CombatResult(
      attackerLosses: attackerLosses, defenderLosses: defenderLosses);
}

/// Validate an attack action. Throws [ArgumentError] on invalid attack.
void validateAttack(
    GameState state, MapGraph mapGraph, AttackAction action, int playerIndex) {
  final sourceTs = state.territories[action.source]!;
  final targetTs = state.territories[action.target]!;

  if (sourceTs.owner != playerIndex) {
    throw ArgumentError(
        'Player $playerIndex does not own source territory \'${action.source}\'');
  }

  if (targetTs.owner == playerIndex) {
    throw ArgumentError(
        'Cannot attack own/friendly territory \'${action.target}\'');
  }

  if (!mapGraph.areAdjacent(action.source, action.target)) {
    throw ArgumentError(
        'Territories \'${action.source}\' and \'${action.target}\' are not adjacent');
  }

  if (sourceTs.armies < action.numDice + 1) {
    throw ArgumentError(
        'Source \'${action.source}\' has ${sourceTs.armies} armies but needs '
        'at least ${action.numDice + 1} to attack with ${action.numDice} dice');
  }
}

/// Execute a single attack round.
///
/// Returns (newState, result, conquered).
///
/// [armiesToMove]: armies to advance into conquered territory.
/// Defaults to action.numDice (minimum per Risk rules).
(GameState, CombatResult, bool) executeAttack(
  GameState state,
  MapGraph mapGraph,
  AttackAction action,
  int playerIndex,
  Random rng, {
  int? defenderDice,
  int? armiesToMove,
}) {
  validateAttack(state, mapGraph, action, playerIndex);

  final targetTs = state.territories[action.target]!;
  final resolvedDefenderDice = defenderDice ?? (targetTs.armies < 2 ? 1 : 2);

  final result = resolveCombat(action.numDice, resolvedDefenderDice, rng);

  final sourceTs = state.territories[action.source]!;
  final newTerritories =
      Map<String, TerritoryState>.of(state.territories);

  int newSourceArmies = sourceTs.armies - result.attackerLosses;
  int newTargetArmies = targetTs.armies - result.defenderLosses;

  final conquered = newTargetArmies <= 0;

  if (conquered) {
    final armiesMoved = armiesToMove ?? action.numDice;
    newSourceArmies -= armiesMoved;
    newTerritories[action.source] =
        TerritoryState(owner: sourceTs.owner, armies: newSourceArmies);
    newTerritories[action.target] =
        TerritoryState(owner: playerIndex, armies: armiesMoved);
  } else {
    newTerritories[action.source] =
        TerritoryState(owner: sourceTs.owner, armies: newSourceArmies);
    newTerritories[action.target] =
        TerritoryState(owner: targetTs.owner, armies: newTargetArmies);
  }

  final newState = conquered
      ? state.copyWith(
          territories: newTerritories, conqueredThisTurn: true)
      : state.copyWith(territories: newTerritories);

  return (newState, result, conquered);
}

/// Auto-resolve attack: loop until conquest or attacker has 1 army.
///
/// Returns (newState, allResults, conquered).
(GameState, List<CombatResult>, bool) executeBlitz(
  GameState state,
  MapGraph mapGraph,
  BlitzAction action,
  int playerIndex,
  Random rng,
) {
  // Validate basic preconditions
  final sourceTs = state.territories[action.source]!;
  final targetTs = state.territories[action.target]!;

  if (sourceTs.owner != playerIndex) {
    throw ArgumentError(
        'Player $playerIndex does not own source territory \'${action.source}\'');
  }
  if (targetTs.owner == playerIndex) {
    throw ArgumentError(
        'Cannot attack own/friendly territory \'${action.target}\'');
  }
  if (!mapGraph.areAdjacent(action.source, action.target)) {
    throw ArgumentError(
        'Territories \'${action.source}\' and \'${action.target}\' are not adjacent');
  }
  if (sourceTs.armies < 2) {
    throw ArgumentError(
        'Source \'${action.source}\' needs at least 2 armies to attack');
  }

  GameState currentState = state;
  final allResults = <CombatResult>[];

  while (true) {
    final src = currentState.territories[action.source]!;
    final tgt = currentState.territories[action.target]!;

    final attackerDice = src.armies - 1 < 3 ? src.armies - 1 : 3;
    final defenderDice = tgt.armies < 2 ? tgt.armies : 2;

    if (attackerDice < 1) break;

    final attackAction = AttackAction(
        source: action.source,
        target: action.target,
        numDice: attackerDice);

    final (newState, result, conquered) = executeAttack(
        currentState, mapGraph, attackAction, playerIndex, rng,
        defenderDice: defenderDice);

    currentState = newState;
    allResults.add(result);

    if (conquered) {
      return (currentState, allResults, true);
    }

    if (currentState.territories[action.source]!.armies <= 1) {
      return (currentState, allResults, false);
    }
  }

  return (currentState, allResults, false);
}
